### PR TITLE
MINOR: Move Timer/TimingWheel to server-common

### PR DESCRIPTION
**Pull request by dajac**
This patch rewrite `Timer` and the related classes in Java and moves them to `server-common` module. It is basically a one to one rewrite of the Scala code. Note that `MockTimer` is not moved as part of this patch. It will be done separately.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: maybe add an additional implements autoCloseable where we can call shutdown?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:65]: we are using this pattern of locking very often now (also used in TS related code for leader epoch snapshot). Could we perhaps extract this out into a utility which takes incoming lambda and executes it in a lock, similar to https://github.com/apache/kafka/blob/513e1c641d63c5e15144f9fcdafa1b56c5e5ba09/core/src/main/scala/kafka/utils/CoreUtils.scala#L182

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:29]: you can move `synchronized` to the method signature above?
```
public synchronized a() {
}
```
is equivalent to 
```
public a() {
   synchronized(this) {
   }
}
```

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:20]: Perhaps we can use AtomicReference here and remove the synchronized blocks since we just need to protect concurrent access to timerTaskEntry?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: I am not sure to understand the benefits of doing this here. My understanding is that `AutoClosable` just adds a `close` method but you still need to class it explicitly or implicitly by using `try-with-resources`. If we want to do this, we would need to bring this change to the `Timer` interface instead of doing it here. I think that this is something that we could consider separately.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:65]: Personally, I am not a big fan of those helper methods for locks in our Java code. I think that we could consider this separately.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:29]: I tried this but spotbugs does not like it because of the access `timerTaskEntry` without synchronisation later on. This is why I kept it like this.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:20]: There is a side effect as well when we mutate `timerTaskEntry`. We call `timerTaskEntry.remove()`. Is it safe to use an `AtomicReference` in this case?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:65]: ok

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: Many automatic bug finders will flag a warning if an object of class which is marked as AutoCloseable but is not closed properly in the code. That is one advantage of using this interface that I find very useful.

Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: Doing is separately is fine for me.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:20]: I was thinking of something like below which would allow remove to happen outside of the lock (and hence won't block the other threads) which should be fine because TimeTaskList#remove() already performs remove with a lock on specific entry.

With AtomicReference, you will write this code as:
```
TimeTaskEntry oldtaskEntry = timerTaskEntry.getAndSet(null)
if (oldtaskEntry != null)
    oldtaskEntry.remove()
```

Similarily for setTimerEntry(), the code could we written as:
```
TimeTaskEntry oldtaskEntry = timerTaskEntry.getAndSet(newEntry)
if (oldtaskEntry != null && !oldtaskEntry.equals(newEntry)) {
    oldtaskEntry.remove()
}
```

Nevertheless, I would understand if we want to keep this PR similar to scala code to avoid any logic changes. Optimizations suggested here can be done in separate PRs later.

Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java:20]: Yeah, that should work. I think that it is preferable to do this separately though. My goal was to have a 1:1 mapping between the scala and the java code in this PR.

**Comment by dajac** [comment]
@divijvaidya As you started reviewing it, are you interested in finishing the review? That would unblock me.

**Comment by divijvaidya** [comment]
Hi @dajac , yes let me take a look first thing tomorrow.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java]: private final (corresponding to val instead of var in scala)

This is only set in the constructor.

(same for other members of this class)

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java]: The timeunit is MILLISECONDS in the scala code here. Is the change to MICROSECONDS intentional? If no, then why didn't a unit test fail? Could we add that missing unit test please?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:36]: does `this` reference exist even when constructor invocation isn't complete?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:19]: Missing compareTo method.

A unit test should have caught this. Please add a unit test if we are having a gap in testing.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java
Comment (divijvaidya) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java:156]: should this be `double` instead of long?

File: core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala
Comment (divijvaidya) [inline core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala:28]: why did we remove the inheritance from CompletableFuture?

**Review by divijvaidya — COMMENTED** [review]
Reviewed the non test files so far. Leaving early feedback while I look at the test files.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: Created a JIRA for this https://issues.apache.org/jira/browse/KAFKA-15085 
It will be a good task for new comers to enter Kafka code base.

File: core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala:27]: This can be removed after we add compare method to TimerTaskEntry

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskList.java]: It is not... Let me fix this.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:36]: I think that it does.

File: core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala
Comment (dajac) [reply core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala:28]: `TimerTask` is an abstract class in Java vs a trait in Scala and we can't extend two classes. Therefore, I have refactored this piece to only extend `TimerTask`.

**Review by divijvaidya — COMMENTED** [review]
Test files look good. One minor comment.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:19]: You meant `compare`, right? We can't implement in Java because it comes from Scala trait. It was only used by `MockTimer` and I have refactored it to not require it any more.

File: core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala
Comment (dajac) [reply core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala:27]: I cannot use `Ordering` in Java. This is why I change it this way.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java:156]: hum... i think that dividing a long by a long gives a long, no?

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (dajac) [inline server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java]: @divijvaidya I was wondering if using `equals` is correct here. It was like this in Scala but as we compare references in all other places, it is a bit weird. Practically, it does not make any difference because `equals` is not overridden anyway.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/SystemTimer.java:29]: Thanks.

**Comment by dajac** [comment]
@divijvaidya Thanks for your review. I have addressed all your comments.

File: core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala
Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala:27]: The Java equivalent of scala `Ordering` could be `Comparable`. That will force you to implement a compareTo function for TimerTaskEntry.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimingWheel.java:156]: you are right, I was thinking from a perspective if this needs to be double for later use. But seems like we will take the floor of the double anyways to compute the bucket which is equivalent to boxing into long.

Hence, long is good here.

File: core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala
Comment (divijvaidya) [reply core/src/main/scala/kafka/raft/TimingWheelExpirationService.scala:28]: ok

File: core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala
Comment (dajac) [reply core/src/test/scala/unit/kafka/utils/timer/MockTimer.scala:27]: Right but I don’t think that the scala priority queue accepts this. This is why I do it inline here.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:19]: By extending `Ordered` in scala for TimerTaskEntry, we wanted to specify that objects of this class are comparable (i.e. functions such as sprt can be used) and should be compared based on their expirationMs. We should implement a similar mechanism in Java using Comparable over here https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (divijvaidya) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java]: yes, it should be a reference comparison. There can be two entries in the list which are same i.e. have same expirationMs and other members but task could be held by only one of them.

Good observation here.

**Review by divijvaidya — COMMENTED** [review]
Looks good to me! The code is consistent with the scala implementation. I haven't reviewed for logical bugs but that is out of scope of this PR.

(waiting for CI test run to complete before approving)

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTaskEntry.java:19]: As explained in the other comment, this is not required any more.

File: server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java
Comment (dajac) [reply server-common/src/main/java/org/apache/kafka/server/util/timer/TimerTask.java]: Fixed in last commit.

**Comment by dajac** [comment]
Failed tests seem unrelated:
```
Build / JDK 11 and Scala 2.13 / testReplication() – org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest
1m 45s
Build / JDK 11 and Scala 2.13 / testOffsetTranslationBehindReplicationFlow() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
1m 2s
Build / JDK 11 and Scala 2.13 / testRestartReplication() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest
3m 8s
Build / JDK 11 and Scala 2.13 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest
1m 29s
Build / JDK 11 and Scala 2.13 / testDescribeAtMinIsrPartitions(String).quorum=zk – kafka.admin.TopicCommandIntegrationTest
6s
Build / JDK 11 and Scala 2.13 / [1] Type=ZK, Name=testNewAndChangedTopicsInDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
14s
Build / JDK 8 and Scala 2.12 / testBumpTransactionalEpoch(String).quorum=kraft – kafka.api.TransactionsTest
1m 20s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testNewAndChangedTopicsInDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
15s
Build / JDK 17 and Scala 2.13 / testBalancePartitionLeaders() – org.apache.kafka.controller.QuorumControllerTest
13s
Build / JDK 17 and Scala 2.13 / shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once_v2] – org.apache.kafka.streams.integration.EosIntegrationTest
```

**Review by divijvaidya — APPROVED** [review]

**Pull request merged**
